datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -188,37 +181,61 @@ impl LogicalPlanBuilder {
                     n_cols
                 );
             }
-            field_types = row
-                .iter()
-                .enumerate()
-                .map(|(j, expr)| {
-                    if let Expr::Literal(ScalarValue::Null) = expr {
-                        nulls.push((i, j));
-                        Ok(field_types[j].clone())
-                    } else {
-                        let data_type = expr.get_type(&empty_schema)?;
-                        if let Some(prev_data_type) = &field_types[j] {
-                            if prev_data_type != &data_type {
-                                return plan_err!("Inconsistent data type across values list at row {i} column {j}. Was {prev_data_type} but found {data_type}")
+        }
+
+        let empty_schema = DFSchema::empty();
+        let mut field_types: Vec<DataType> = Vec::with_capacity(n_cols);
+        for j in 0..n_cols {
+            let mut common_type: Option<DataType> = None;
+            for (i, row) in values.iter().enumerate() {
+                let value = &row[j];
+                if let Expr::Literal(ScalarValue::Null) = value {
+                    continue;
+                } else {
+                    let data_type = value.get_type(&empty_schema)?;
+                    if let Some(prev_type) = common_type {
+                        // get common type of each row values.
+                        match comparison_coercion(&data_type, &prev_type) {
```

Comment:
The concern is that I'm not sure `comparison_coercion` is suitable for constructing Value, it is easily misused `coercion`. I would recommend we start by adding numeric coercion for example the integer to floating. We can add more and more coercion rules covered with the test. If eventually, it is the same as `comparison_coercion`, then we can replace it.  But, I think we should start with simple.

Reply:
Agree with @jayzhan211 
Maybe `comparison_binary_numeric_coercion` can be started with?

Reply:
Using `comparison_binary_numeric_coercion` doesn't work, it causes errors with Utf8 values, I've added a `values_coercion` function, please check if it's appropriate? @jayzhan211 @yyy1000

Reply:
Pull string coercion part out of `string_coercion` as `pure_string_coercion`
```
        (Utf8, Utf8) => Some(Utf8),
        (LargeUtf8, Utf8) => Some(LargeUtf8),
        (Utf8, LargeUtf8) => Some(LargeUtf8),
        (LargeUtf8, LargeUtf8) => Some(LargeUtf8),
```

Reply:
I had once tried to introduce type resolution coercion in https://github.com/apache/datafusion/pull/10268, but not yet merged.

Reply:
In `string_coercion` function there are coercion rules for check `List` types, this logic seems incorrect, we can move the `List` type checks out of `string_coercion` and write a separate function `list_coercion`, is this correct?

Reply:
I agree the `List coercion` is not correct. If you can remove it without failure then it is nice to remove it.